ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -700,6 +771,12 @@ VALUE Message_decode(VALUE klass, VALUE data) {
 
   stackenv_uninit(&se);
 
+  if (desc->has_udts_transitively) {
```

Comment:
Imagine the case where no repeated fields are present (this matters for a reason I'll describe below). We could see a nested chain of start-submessage tags that sets a deeply nested field, then some completely unrelated data, then some appended data that goes down the same nested chain and alters or overwrites the submessage at the end of the long chain. Because of this, we can't convert *any* submessage at the end of such a chain to its user type until we've seen the entire stream of bytes.

Because repeated fields have a "tag seen again" behavior of "append new submessage" rather than "alter existing submessage", we *can* parse immediately if there's a repeated field anywhere in the field path from root. I haven't implemented that optimization yet.